### PR TITLE
support for image per song

### PR DESCRIPTION
**Pull request by Greo83**
Optionally save and show the picture of a song related to the song instead of its album. Use this feature if your music collection contains more single songs than full albums.

Original discussions:
[https://github.com/ampache/ampache/pull/2618](url)
[https://github.com/ampache/ampache/pull/1847](url)

### CODE REVIEW

**Comment by wagnered** [comment]
Almost.  You forgot to run `php-cs-fixer -vvv fix` before pushing to your fork.☺️

Otherwise it looks quite clean.

Here is a ref for the proper way to [sync your fork](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/syncing-a-fork).

I'll give your code a try as soon as I can get to a stopping point in my current project.

**Comment by lachlan-00** [comment]
Something that is hard to remember is to run the codestyle checkers. (i wish you could script that as part of a commit)
```
./php-cs-fixer -v fix && ./phpcbf --standard=.phpcs.xml lib/class
```

I'll be able to work on this during the week, it's been a rough house week here.

**Comment by wagnered** [comment]
So far so good. Now how do I test it?
Really need more explanation. 

Are the images dispayed on the song page of the web UI?  What happens to the existing cover/artist images in the db?

  Do I need to have non artist/cover images embedded in the file tag?  I ask this because it affect the updating onf the tags which must be written in it's entirety each time.

File: lib/class/art.class.php [hunk 1]
```diff
@@ -948,8 +948,17 @@ public function generate_thumb($image, $size, $mime)
      * @param string $type
      * @return string
      */
-    public static function get_from_source($data, $type = 'album')
```
Comment (wagnered) [inline lib/class/art.class.php:951]: why did you remove the default value for parameter 'type'?

File: arts.php [hunk 2]
```diff
@@ -160,7 +160,12 @@
         // Prevent the script from timing out
         set_time_limit(0);
 
-        $image      = Art::get_from_source($_SESSION['form']['images'][$image_id], 'album');
+        $show_song_art = AmpConfig::get('show_song_art') ?: false;
```
Comment (wagnered) [inline arts.php:163]: The ternary operatioin `?:` isn't necessary and also isn't fool proof.  If you look at the function definition ` AmpConfig::get('show_song_art')` it has a second parameter  which can set a return value.

Just a reminder for the next PR you do.

File: lib/class/art.class.php (hunk 1 above)
Comment (lachlan-00) [reply lib/class/art.class.php:951]: did a search on it and don't see any calls with a null type, but it's still a fallback i would keep

**Comment by Greo83** [comment]
> So far so good. Now how do I test it?
> Really need more explanation.
> 
> Are the images dispayed on the song page of the web UI? What happens to the existing cover/artist images in the db?
> 
> Do I need to have non artist/cover images embedded in the file tag? I ask this because it affect the updating onf the tags which must be written in it's entirety each time.

Imagine you have three songs/files. Standalone songs. Each song has one picture in its ID tag. In fact the picture shows the title of the song, so it is absolutely unique. The album tag is empty.
Ampache's standard behaviour in this scenario then is: When importing the first file it creates a new album entry called "orphaned". This new album gets the picture of the first song. The next files are attached to the same album - "orphaned". The pictures are not imported because the album already exists.
When the songs are shown in Ampache, no matter if in the web player or playlist or whatever, all three songs show the picture of the first file. This is because all of them are linked to the "orphaned" album and Ampache always shows the picture of the album.

When I started to investigate this I had the idea to handle the album "orphaned" in a different way. But then I found out that all pictures are linked to the albums in the SQL database. 
Because of this I created the option to additionally (!) add the picture with a relation to song. For our scenario above this means: You still have the album "orphaned" with the first picture, everything as before. But additionally the picture of each file is also saved in the database, with the relation "song". When playing a song, showing a playlist and so on Ampache now shows the song-related picture instead of the album-related picture.

This feature is interesting for people who have mostly stand-alone songs instead of full albums in their library.

File: lib/class/art.class.php (hunk 1 above)
Comment (Greo83) [reply lib/class/art.class.php:951]: My problem was that the function always used the default value 'album', although I called the function with the parameter value 'song'. My parameter was ignored, it always was overwritten by 'album'. I had a long debugging session until I realized this.

But to be honest I'm absolutely new to PHP - maybe the reason was different?

Comment (lachlan-00) [reply lib/class/art.class.php:951]: i'll check it out, it's possible that something is being sent null

Comment (wagnered) [reply lib/class/art.class.php:951]: Well here is the [php authority](https://www.php.net/manual/en/).  It's always good to have a default value: no one knows how it migt be accessed in the future.

File: image.php [hunk 3]
```diff
@@ -51,7 +51,7 @@
 
 // FIXME: Legacy stuff - should be removed after a version or so
 if (!filter_has_var(INPUT_GET, 'object_type')) {
-    $_GET['object_type'] = 'album';
```
Comment (lachlan-00) [inline image.php:54]: what does this change do? i'm assuming it's setting a default type for art?

File: arts.php (hunk 2 above)
Comment (lachlan-00) [reply arts.php]: these are changed into different operators based on need now, the false isn't required in the way it's written

File: lib/class/art.class.php (hunk 1 above)
Comment (lachlan-00) [reply lib/class/art.class.php:951]: There's code below to set album/song based on the config. i'm not sure when this would ever fire but there must be something not format or set before the call somewhere outside this
```
public static function get_from_source($data, $type)
    {
        if (!isset($type)) {
            $type = (AmpConfig::get('show_song_art')) ? 'song' : 'album';
        }
```

**Comment by wagnered** [comment]
>  Each song has one picture in its ID tag. In fact the picture shows the title of the song, so it is absolutely unique. The album tag is empty.

How are those images added to the tags?

What makes an image unique is a combination of the assigned picture's id as denoted by id3.org and the description of the image.  It matters not if an "album" frame exists.  The picture id associates the image with the cover, artist, conductor.  Granted, Ampache does a poor job of differentiating between artist an cover.

I'm concerned about the writing of the images back to the file.  The getID3 library adheres to the id3.org specs for id3v2.3 which  Ampache uses. If the embedded images don't conform, they will be lost upon tag rewrite. 

I think what you want can be done more transparently by using the `kind` column. Another config switch might not be necessary.  There are so any now that it take an undergrad degree to fully understand Ampache 😄.

I would like you to give it some thought.

File: image.php (hunk 3 above)
Comment (Greo83) [reply image.php:54]: To be honest I don't remember. I changed that line 1 1/2 years ago in a lot of debugging and try and error sessions. When I review that now I would agree. I think this default setting should not be hard coded, but depending on the config file.

**Comment by Greo83** [comment]
> How are those images added to the tags?
> 

What do you mean with that? I checked some files. The image type of the ID tag is "front cover" mostly.

> I'm concerned about the writing of the images back to the file. The getID3 library adheres to the id3.org specs for id3v2.3 which Ampache uses. If the embedded images don't conform, they will be lost upon tag rewrite.

Okay, I didn't try that. I use Ampache as a read-only tool. Is the object_type field really passed back directly into the ID tag?

>
> I think what you want can be done more transparently by using the `kind` column. 

Theoretically it should work, yes. But then all the database statements that Ampache calls have to be changed. This seems to be more effort, isn't it? Additionally, I think from a database point of view this is not correct, because the image is assigned to the song - not to the album. As far as I understood all the SQL and PHP code around the object_type field everything was designed to attach images also to different other tables than the album table. It just has not been developed yet.

**Comment by wagnered** [comment]
The fact that you want to use cover art (id=3) generically  does complicate the issue.  I want the spec to be that "song"art could have any ID except that of cover and artist.  This will make the design  much less complicated.  The ID of your current embedded images coud be altered relatively easily with easy tag or eyeD3 on Linux.  I'm placing this limitation on the image spec because I shall be implementing image display on the song page and perhaps the song row page if I can squeeze it on.    Your PR  doesn't do this. 

>  But then all the database statements that Ampache calls have to be changed.

Not necessarily.  But  familiarity with a project in total helps one see solutions that affect the current development as well as future possibilities.  

But the near-future architecture will include a database managent system (dbms) that will make SQL statements  more manageable.

File: lib/class/catalog.class.php [hunk 4]
```diff
@@ -1460,14 +1460,22 @@ public function gather_art($songs = null, $videos = null)
...
+            }
         } else {
             $searches['album']  = array();
             $searches['artist'] = array();
+            if ($gather_song_art) {
+                $searches['song'] = array();
+            }
             foreach ($songs as $song_id) {
                 $song = new Song($song_id);
                 if ($song->id) {
```
Comment (wagnered) [inline lib/class/catalog.class.php:1480]: The objects were already created/returned on line 1471. To be consistent, I suggest that song IDs be returned and objects created here.

**Comment by wagnered** [comment]
I'm feeling more comfortable with the PR because it seems to apply only to API access

I think that because Art::gather cycles through every entry in `art_order` there is an assumption that the `art_order` setting includes only "db" and "tags".  How would adding spotify to the list affect the results?

**Comment by AshotN** [comment]
> Something that is hard to remember is to run the codestyle checkers. (i wish you could script that as part of a commit)
> 
> ```
> ./php-cs-fixer -v fix && ./phpcbf --standard=.phpcs.xml lib/class
> ```
> 
> I'll be able to work on this during the week, it's been a rough house week here.

Doesn’t GitHub actions do this?

**Comment by lachlan-00** [comment]
> 
> 
> > Something that is hard to remember is to run the codestyle checkers. (i wish you could script that as part of a commit)
> > ```
> > ./php-cs-fixer -v fix && ./phpcbf --standard=.phpcs.xml lib/class
> > ```
> > 
> > 
> > I'll be able to work on this during the week, it's been a rough house week here.
> 
> Doesn’t GitHub actions do this?

it looks like it might actually... i'll check it out.
image

**Comment by lachlan-00** [comment]
@Greo83 i'm going to start going through this to merge now.

**Comment by usox** [comment]
> > Doesn’t GitHub actions do this?
> 
> it looks like it might actually... i'll check it out.

I could provide working workflow configs/examples if you wish.

**Comment by lachlan-00** [comment]
> > > Doesn’t GitHub actions do this?
> > 
> > it looks like it might actually... i'll check it out.
> 
> I could provide working workflow configs/examples if you wish.

I think this would help docker as well because I could build a zip/gzip for building the image for develop

**Comment by AshotN** [comment]
Correct me if I’m wrong. But my understanding is this will add a image for every song. Wouldn’t an image override option be easier? 

So if the album has an image, but a specific song within that album has a different image, that different art alone would be stored. 

To avoid needless duplication.

**Comment by usox** [comment]
> > > > Doesn’t GitHub actions do this?
> > > 
> > > 
> > > it looks like it might actually... i'll check it out.
> > 
> > 
> > I could provide working workflow configs/examples if you wish.
> 
> I think this would help docker as well because I could build a zip/gzip for building the image for develop

Oh, I guess I got you wrong. You're talking about automatically fixing the code after pushing a commit (serverside at github), right?
I'm not a big fan of automatically altering peoples commits. A failing build of a PR shouldn't be a big problem for the developers here (My first PR failed, too). Maybe, we could add a CONTRIBUTING.md at root level of the repo to better communicate our rules...

**Comment by AshotN** [comment]
> > > > > Doesn’t GitHub actions do this?
> > > > 
> > > > 
> > > > it looks like it might actually... i'll check it out.
> > > 
> > > 
> > > I could provide working workflow configs/examples if you wish.
> > 
> > 
> > I think this would help docker as well because I could build a zip/gzip for building the image for develop
> 
> Oh, I guess I got you wrong. You're talking about automatically fixing the code after pushing a commit (serverside at github), right?
> I'm not a big fan of automatically altering peoples commits. A failing build of a PR shouldn't be a big problem for the developers here (My first PR failed, too). Maybe, we could add a CONTRIBUTING.md at root level of the repo to better communicate our rules...

Just set it up to fail instead of modifying the commit. But for formatting it shouldn’t really be an issue.

**Comment by kuzi-moto** [comment]
> > > > > Doesn’t GitHub actions do this?
> > > > 
> > > > 
> > > > it looks like it might actually... i'll check it out.
> > > 
> > > 
> > > I could provide working workflow configs/examples if you wish.
> > 
> > 
> > I think this would help docker as well because I could build a zip/gzip for building the image for develop
> 
> Oh, I guess I got you wrong. You're talking about automatically fixing the code after pushing a commit (serverside at github), right?
> I'm not a big fan of automatically altering peoples commits. A failing build of a PR shouldn't be a big problem for the developers here (My first PR failed, too). Maybe, we could add a CONTRIBUTING.md at root level of the repo to better communicate our rules...

I swear we had it somewhere. Obviously it's not easy to find, so it definitely should exist in the CONTRIBUTING doc.

**Comment by lachlan-00** [comment]
k, happy with this now, will merge once it's built in travis

**Comment by lachlan-00** [comment]
i migrated to travis.com from the org so i'm not waiting for that.
